virttest/qemu_vm.py

```diff
@@ -3708,134 +3641,270 @@ def create(
+        All parameters are optional. If name, params or root_dir are not
+        supplied, the respective values stored as class attributes are used.
+
+        :param name: The name of the object
+        :param params: A dict containing VM params
+        :param root_dir: Base directory for relative filenames
+        :param timeout: The timeout value for creating monitors
+        :param migration_mode: If supplied, start VM for incoming migration
+                using this protocol (either 'rdma', 'x-rdma', 'rdma', 'tcp', 'unix' or 'exec')
+        :param migration_exec_cmd: Command to embed in '-incoming "exec: ..."'
+                (e.g. 'gzip -c -d filename') if migration_mode is 'exec'
+                default to listening on a random TCP port
+        :param migration_fd: Open descriptor from machine should migrate.
+        :param mac_source: A VM object from which to copy MAC addresses. If not
+                specified, new addresses will be generated.
 
-            # Update mac and IP info for assigned device
-            # NeedFix: Can we find another way to get guest ip?
-            if params.get("mac_changeable") == "yes":
-                utils_net.update_mac_ip_address(self)
+        :raise VMCreateError: If qemu terminates unexpectedly
+        :raise VMKVMInitError: If KVM initialization fails
+        :raise VMHugePageError: If hugepage initialization fails
+        :raise VMImageMissingError: If a CD image is missing
+        :raise VMHashMismatchError: If a CD image hash has doesn't match the
+                expected hash
+        :raise VMBadPATypeError: If an unsupported PCI assignment type is
+                requested
+        :raise VMPAError: If no PCI assignable devices could be assigned
+        :raise TAPCreationError: If fail to create tap fd
+        :raise BRAddIfError: If fail to add a tap to a bridge
+        :raise TAPBringUpError: If fail to bring up a tap
+        :raise PrivateBridgeError: If fail to bring the private bridge
+        """
+        if name is not None:
+            self.name = name
+            self.devices = None  # Representation changed
+        if params is not None:
+            self.params = params
+            self.devices = None  # Representation changed
+        if root_dir is not None:
+            self.root_dir = root_dir
+            self.devices = None  # Representation changed
+        name = self.name
+        params = self.params
+        root_dir = self.root_dir
 
+        # Make sure the following code is not executed by more than one thread
+        # at the same time
+        lockfile = open(CREATE_LOCK_FILENAME, "w+")
+        fcntl.lockf(lockfile, fcntl.LOCK_EX)
+
+        try:
+            self.define(name, params, root_dir, mac_source)
+            # Add migration command line if required
+            if migration_mode:
+                self.add_incoming(migration_mode, migration_exec_cmd, migration_fd)
+            self.start(timeout)
```

Comment:
There is one concern from the upstream scenario that the user will use the previous job's VM to run the next job by calling `vm.create()`  without `name`, `params`, and `root_dir`, it means that we do not need to define the VM again and then start the VM with the pre-defined VM devices.
The root cause is that the `vm.create`  integrates inappropriate functionality: `re-create VM`.
However, what we can do is just keep the older design to avoid regression in the future.
Please let me know your opinions.

Reply:
Thank you @YongxueHong. With this scenario, I would like to cut the following code from make_create_command and move to `def create`, to let `define` and `make_create_command` focus on their operation.
```python
        if (
            name is None
            and params is None
            and root_dir is None
            and self.devices is not None
        ):
            return self.devices, self.spice_options
```

Reply:
Thank you for pointing it out. I use this checkpoint to avoid the issue.
https://github.com/avocado-framework/avocado-vt/pull/3761/files#diff-274c35be00ea6170de3ec316e8ed20eca958a5de150a49770287fd1852a4e8feR3365-R3370
https://github.com/avocado-framework/avocado-vt/pull/3761/files#diff-274c35be00ea6170de3ec316e8ed20eca958a5de150a49770287fd1852a4e8feR3387-R3388